packages/ra-ui-materialui/src/button/SaveButton.spec.tsx

```diff
@@ -337,4 +337,52 @@ describe('<SaveButton />', () => {
             });
         });
     });
+
+    it('should disable <SaveButton/> if an input is being validated asynchronously', async () => {
+        const dataProvider = ({
+            getOne: () =>
+                Promise.resolve({
+                    data: { id: 123, title: 'lorem' },
+                }),
+            update: (_, { data }) => Promise.resolve({ data }),
+        } as unknown) as DataProvider;
+
+        const validateAgainstServer = value =>
+            new Promise(resolve => setTimeout(() => value, 400));
+
+        const validateAsync = async (value, allValues) => {
+            const isUnique = await validateAgainstServer(value);
+            if (!isUnique) {
+                return 'Already used!';
+            }
+            return undefined;
+        };
+
+        const { queryByDisplayValue, getByLabelText } = renderWithRedux(
+            <DataProviderContext.Provider value={dataProvider}>
+                <ThemeProvider theme={theme}>
+                    <Edit {...defaultEditProps}>
+                        <SimpleForm>
+                            <TextInput
+                                source="title"
+                                validate={validateAsync}
+                            />
+                        </SimpleForm>
+                    </Edit>
+                </ThemeProvider>
+            </DataProviderContext.Provider>,
+            { admin: { resources: { posts: { data: {} } } } }
+        );
+        // waitFor for the dataProvider.getOne() return
+        await waitFor(() => {
+            expect(queryByDisplayValue('lorem')).toBeDefined();
+        });
+
+        // change one input to enable the SaveButton (which is disabled when the form is pristine)
+        fireEvent.change(getByLabelText('resources.posts.fields.title'), {
+            target: { value: 'ipsum' },
+        });
+
+        expect(getByLabelText('ra.action.save')['disabled']).toEqual(true);
```

Comment:
I'm not sure you're testing the right thing. 

The SaveButton will be disabled at start because no value was changed. Then, it will be disabled because the validation is ongoing. Finally, when the validation has finished, it should either be enabled (if the validation is ok) or disabled (if the validation has failed).

Your test should reflect that and test all cases.

Reply:
Actually during async validation the button should be disabled, after that should be enabled again, unless the form is pristine